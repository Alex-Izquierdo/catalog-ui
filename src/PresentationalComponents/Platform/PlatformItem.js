import React from 'react';
import { withRouter } from 'react-router-dom';
import { connect } from 'react-redux';
import './platformitem.scss';
import propTypes from 'prop-types';
import CatItemSvg from '../../assets/images/vendor-openshift.svg';
import ImageWithDefault from '../Shared/ImageWithDefault';
import { PlatformStore, fetchPlatformItems } from '../../Store/Actions/PlatformActions';
import { hideModal, showModal } from '../../Store/Actions/MainModalActions';
import { GridItem, Card, CardHeader, CardBody, CardFooter, Checkbox } from '@patternfly/react-core';
import { Dropdown, DropdownItem, DropdownPosition, DropdownToggle } from '@patternfly/react-core';
import itemDetails from '../../PresentationalComponents/Shared/CardCommon';
import { bindMethods, consoleLog } from '../../Helpers/Shared/Helper';
import CardCheckbox from '../Shared/CardCheckbox';

const TO_DISPLAY = [ 'description' ];

const mapDispatchToProps = dispatch => {
    return {
        hideModal: () => dispatch(hideModal()),
        showModal: (modalProps, modalType) => {
            dispatch(showModal({ modalProps, modalType }));
        }
    };
};

class PlatformItem extends React.Component {
    constructor(props) {
        super(props);
        this.state = { isOpen: true, showMenu: false, isEditMode: false };
    };

    onSelect = (event) => {
        consoleLog('This is the selected state:', this.state);
        consoleLog('This is the selected event:', event);

        this.props.showModal({
            open: true,
            itemdata: this.props,
            closeModal: this.props.hideModal
        }, 'addportfolio');

        this.setState({
            ...this.state,
            isOpen: !this.state.isOpen
        });
    };

    isChecked = () => {
        if (this.props.isChecked)
        {return this.props.isChecked(this.props.id);}
        else
        {return false;}
    }

    render() {
        return (
            <GridItem key={ this.props.id } GridItem sm={ 6 } md={ 4 } lg={ 4 } xl={ 3 }>
                <Card key={ this.props.id }>
                    <CardHeader className="card_header">
                        <ImageWithDefault src={ this.props.imageUrl || CatItemSvg } defaultSrc={ CatItemSvg } width="50" height="50" />
                        { this.props.isEditMode && <CardCheckbox id={ this.props.id } checked={ this.isChecked } onChange={ this.props.onCheckboxClick } /> }
                    </CardHeader>
                    <CardBody className="card_body">
                        <h4>{ this.props.name }</h4>
                        { itemDetails(this.props, TO_DISPLAY) }
                    </CardBody>
                    <CardFooter/>
                </Card>
            </GridItem>
        );
    };
}

PlatformItem.propTypes = {
    history: propTypes.object,
    showModal: propTypes.func,
    hideModal: propTypes.func,
    imageUrl: propTypes.string,
    id: propTypes.string,
    name: propTypes.string,
    isEditMode: propTypes.string,
    checkedItems: propTypes.object,
    isChecked: propTypes.string,
    onCheckboxClick: propTypes.func,
    onCheckboxClick: propTypes.func
};

export default withRouter(PlatformItem);

