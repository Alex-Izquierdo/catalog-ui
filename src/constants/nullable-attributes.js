export const PORTFOLIO_ITEM_NULLABLE = [
  'display_name', 'description', 'long_description', 'distributor', 'documentation_url', 'support_url', 'workflow_ref'
];
