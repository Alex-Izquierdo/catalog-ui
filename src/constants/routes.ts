export const ADD_PORTFOLIO_ROUTE = '/portfolios/add-portfolio';
export const ADD_PRODUCTS_ROUTE = '/portfolios/add-products';
export const EDIT_ORDER_PROCESS_ROUTE = '/portfolios/edit-order-process';
export const EDIT_PORTFOLIO_ROUTE = '/portfolios/edit';
export const EDIT_PORTFOLIO_WORKFLOW_ROUTE = '/portfolios/edit-workflow';
export const EDIT_SURVEY_ROUTE = '/portfolios/portfolio-item/edit-survey';
export const NESTED_EDIT_ORDER_PROCESS_ROUTE =
  '/portfolios/portfolio/edit-order-process';
export const NESTED_EDIT_PORTFOLIO_ROUTE =
  '/portfolios/portfolio/edit-portfolio';
export const NESTED_REMOVE_PORTFOLIO_ROUTE =
  '/portfolios/portfolio/remove-portfolio';
export const NESTED_SHARE_PORTFOLIO_ROUTE =
  '/portfolios/portfolio/share-portfolio';
export const NESTED_WORKFLOW_PORTFOLIO_ROUTE =
  '/portfolios/portfolio/edit-workflow';
export const ORDER_ROUTE = '/orders/order';
export const ORDERS_ROUTE = '/orders';
export const PLATFORM_INVENTORIES_ROUTE = '/platforms/platform-inventories';
export const PLATFORM_ROUTE = '/platforms/platform';
export const PLATFORM_SERVICE_OFFERINGS_ROUTE = '/platforms/service-offerings';
export const PLATFORM_TEMPLATES_ROUTE = '/platforms/platform-templates';
export const PLATFORM_DETAILS_ROUTE = '/platforms/platform-details';
export const PLATFORMS_ROUTE = '/platforms';
export const PORTFOLIO_ROUTE = '/portfolios/portfolio';
export const PORTFOLIO_ITEM_ROUTE = '/portfolios/portfolio-item';
export const PORTFOLIO_ITEM_ORDER_ROUTE = '/portfolios/portfolio-item/order';
export const PORTFOLIO_ITEM_ROUTE_EDIT = '/portfolios/portfolio-item/edit';
export const PORTFOLIO_ITEM_EDIT_ORDER_PROCESS_ROUTE =
  '/portfolios/portfolio-item/edit-order-process';
export const PORTFOLIOS_ROUTE = '/portfolios';
export const REMOVE_PORTFOLIO_ROUTE = '/portfolios/remove';
export const SHARE_PORTFOLIO_ROUTE = '/portfolios/share';
export const WORKFLOW_PORTFOLIO_ROUTE = '/portfolios/edit-workflow';
export const ORDER_PROCESSES_ROUTE = '/order-processes';
export const ADD_ORDER_PROCESS_ROUTE = '/order-processes/add-order-process';
export const REMOVE_ORDER_PROCESS_ROUTE = '/order-processes/remove';
export const UPDATE_ORDER_PROCESS_ROUTE = '/order-processes/update';
