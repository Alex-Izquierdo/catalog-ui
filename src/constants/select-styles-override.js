const style = {
  container: provided => ({
    ...provided,
    '&:focus': {
      outline: 'none'
    }
  }),
  control: (provided, { isFocused }) => ({
    ...provided,
    borderRadius: 0,
    borderColor: '#EDEDED',
    backgroundColor: '#FFFFFF',
    borderBottomColor: isFocused ? '#007bba' : '#282d33',
    cursor: 'pointer',
    boxShadow: 'none',
    minHeight: 36,
    width: 256,
    '&:hover': {
      borderBottomColor: '#007bba'
    }
  }),
  indicatorSeparator: () => ({}),
  dropdownIndicator: provided => ({
    ...provided,
    color: '#282d33',
    paddingBottom: 6,
    paddingTop: 6
  }),
  menu: provided => ({
    ...provided,
    borderRadius: 'none',
    boxShadow:
      '0 0.0625rem 0.0625rem 0rem rgba(3,3,3,.05),0 0.25rem 0.5rem 0.25rem rgba(3,3,3,.06)',
    paddingBottom: '0.5rem',
    paddingTop: '0.5rem',
    marginTop: 4,
    width: 'calc(100% - 4px)',
    marginRight: 2,
    marginLeft: 2
  }),
  option: (provided, { isFocused }) => ({
    ...provided,
    cursor: 'pointer',
    textAlign: 'left',
    color: '#282d33',
    fontSize: '1rem',
    fontWeight: '400',
    lineHeight: '1.5',
    paddingBottom: '0.5rem',
    paddingTop: '0.5rem',
    paddingLeft: '1rem',
    paddingRight: '1rem',
    backgroundColor: isFocused ? '#ededed' : 'transparent',
    '&:hover': {
      backgroundColor: '#ededed'
    }
  }),
  placeholder: provided => ({
    ...provided,
    left: 20
  }),
  singleValue: provided => ({
    ...provided,
    left: 20,
    maxWidth: 'calc(100% - 28px)'
  })
};

export default style;
