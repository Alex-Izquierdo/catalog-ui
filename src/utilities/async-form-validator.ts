import awesomeDebouncePromise, {
  AwesomeDebounceOptions
} from 'awesome-debounce-promise';

const asyncFormValidator = (
  asyncFunction: (...args: any[]) => any,
  debounceTime = 250,
  options: Partial<AwesomeDebounceOptions> = { onlyResolvesLast: false }
): any => awesomeDebouncePromise(asyncFunction, debounceTime, options);

export default asyncFormValidator;
