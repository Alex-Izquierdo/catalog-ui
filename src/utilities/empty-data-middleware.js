const emptyDataMiddleware = () => (dispatch) => (action) => {
  const nextAction = { ...action };
  if (
    action.type.match(/_FULFILLED$/) &&
    action.payload &&
    action.meta &&
    action.payload.data &&
    action.payload.meta
  ) {
    nextAction.payload.meta.noData =
      nextAction.payload.meta.count === 0 &&
      nextAction.meta.filter.length === 0;
    return dispatch(nextAction);
  }

  return dispatch(nextAction);
};

export default emptyDataMiddleware;
