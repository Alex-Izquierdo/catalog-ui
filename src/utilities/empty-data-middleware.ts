import { Dispatch } from 'redux';
import { ReduxAction } from '../types/common-types';

const emptyDataMiddleware = () => (dispatch: Dispatch) => (
  action: ReduxAction
): ReduxAction => {
  const nextAction = { ...action };
  if (
    action.type.match(/_FULFILLED$/) &&
    action.payload &&
    action.meta &&
    action.payload.data &&
    action.payload.meta
  ) {
    const noFilter = Object.prototype.hasOwnProperty.call(
      nextAction.meta,
      'filters'
    )
      ? Object.values(nextAction.meta?.filters || {}).every(
          (value?: string) => typeof value === 'undefined' || value.length === 0
        )
      : nextAction.meta?.filter?.length === 0;

    nextAction.payload.meta.noData =
      nextAction.payload.meta.count === 0 && noFilter;
    return dispatch(nextAction);
  }

  return dispatch(nextAction);
};

export default emptyDataMiddleware;
