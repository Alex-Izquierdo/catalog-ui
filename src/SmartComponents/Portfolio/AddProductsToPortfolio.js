import React, { Component } from 'react';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import propTypes from 'prop-types';
import { Section } from '@red-hat-insights/insights-frontend-components';
import { addNotification } from '@red-hat-insights/insights-frontend-components/components/Notifications';
import { fetchPlatformItems } from '../../redux/Actions/PlatformActions';
import ContentGallery from '../../SmartComponents/ContentGallery/ContentGallery';
import MainModal from '../Common/MainModal';
import '../Platform/platformitems.scss';
import '../../SmartComponents/Portfolio/portfolio.scss';
import PortfolioOrderToolbar from '../../PresentationalComponents/Portfolio/PortfolioOrderToolbar';
import AddProductsTitleToolbar from '../../PresentationalComponents/Portfolio/AddProductsTitleToolbar';
import PlatformDashboard from '../../PresentationalComponents/Platform/PlatformDashboard';
import PlatformSelectToolbar from '../../SmartComponents/Common/PlatformSelectToolbar';
import { addToPortfolio } from '../../redux/Actions/PortfolioActions';
import PlatformItem from '../../PresentationalComponents/Platform/PlatformItem';

class AddProductsToPortfolio extends Component {
    state = {
      selectedPlatforms: [],
      checkedItems: []
    };

    async fetchData(platforms) {
      await this.props.fetchPlatformItems(platforms);
    }

    onPlatformSelectionChange = (selectedValues) => {
      this.setState({ selectedPlatforms: selectedValues });
      this.fetchData({ platform: selectedValues[0].id });
    };

    onToggleItemSelect = (event) => {
      const item = event.target.id;
      const isChecked = event.target.checked;
      console.log('item select: ', event);
      if (isChecked) {
        this.setState(state => {
          const checkedItems = [ ...state.checkedItems, item ];
          return {
            checkedItems
          };
        });
      }
      else {
        this.setState(prevState => ({ checkedItems: prevState.checkedItems.filter((value) => { return !(value === item.id);}) }));
      }
    };

    onAddToPortfolio = () =>
      this.props.addToPortfolio(this.props.portfolio.id, this.state.checkedItems)
      .then(() => this.props.resetViewMode(null, true));

    render() {
      let filteredItems = [];
      if (this.props.platformItems) {
        filteredItems = {
          items: this.props.platformItems.map(item => (
            <PlatformItem
              key={ item.id }
              { ...item }
              editMode
              onToggleItemSelect={ this.onToggleItemSelect }
              checkedItems={ this.state.checkedItems }
            />)),
          isLoading: this.props.isLoading
        };
      }

      let title = (this.props.portfolio) ? this.props.portfolio.name : '';
      return (
        <Section>
          <PortfolioOrderToolbar/>
          <AddProductsTitleToolbar title={ title }
            onClickAddToPortfolio = { this.onAddToPortfolio }
            onClickCancelAddProducts={ this.props.resetViewMode }/>
          <PlatformSelectToolbar onOptionSelect={ this.onPlatformSelectionChange } { ...this.props } />
          { (this.state.selectedPlatforms.length > 0) &&
                    this.state.selectedPlatforms.map((platform)=> {return (<ContentGallery key={ platform.id } { ...filteredItems }
                      title={ platform.name }
                      editMode = { true }
                      onToggleSelect = { this.onToggleItemSelect }
                      checkedItems = { this.state.checkedItems }
                    />);}) }
          { (this.state.selectedPlatforms.length < 1) && <PlatformDashboard/> }
          <MainModal/>
        </Section>
      );
    }
}

const mapStateToProps = ({ platformReducer: { platformItems, isPlatformDataLoading }}) => ({
  platformItems,
  isLoading: isPlatformDataLoading
});

const mapDispatchToProps = dispatch => bindActionCreators({
  addNotification,
  fetchPlatformItems,
  addToPortfolio
}, dispatch);

AddProductsToPortfolio.propTypes = {
  platformItems: propTypes.array,
  isLoading: propTypes.bool,
  isEditMode: propTypes.bool,
  addToPortfolio: propTypes.func,
  resetViewMode: propTypes.func,
  fetchPlatformItems: propTypes.func,
  portfolio: propTypes.shape({
    name: propTypes.string,
    id: propTypes.oneOfType([ propTypes.string, propTypes.number ]).isRequired
  }).isRequired
};

export default connect(mapStateToProps, mapDispatchToProps)(AddProductsToPortfolio);
