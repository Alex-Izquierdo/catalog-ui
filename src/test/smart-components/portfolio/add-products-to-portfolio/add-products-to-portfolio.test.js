import React from 'react';
import { act } from 'react-dom/test-utils';
import thunk from 'redux-thunk';
import { rawComponents } from '@data-driven-forms/pf4-component-mapper';
import { Provider } from 'react-redux';
import { mount, shallow } from 'enzyme';;
import { MemoryRouter } from 'react-router-dom';
import { shallowToJson } from 'enzyme-to-json';
import configureStore from 'redux-mock-store' ;
import promiseMiddleware from 'redux-promise-middleware';
import { notificationsMiddleware } from '@redhat-cloud-services/frontend-components-notifications/';

import { TOPOLOGICAL_INVENTORY_API_BASE, CATALOG_API_BASE, SOURCES_API_BASE } from '../../../../utilities/constants';
import PlatformItem from '../../../../presentational-components/platform/platform-item';
import AddProductsToPortfolio from '../../../../smart-components/portfolio/add-products-to-portfolio';

describe('<AddProductsToPortfolio />', () => {
  let initialProps;
  const middlewares = [ thunk, promiseMiddleware, notificationsMiddleware() ];
  let mockStore;
  mockStore = configureStore(middlewares);
  let ComponentWrapper = ({ store, children }) => (
    <Provider store={ store }>
      <MemoryRouter>
        { children }
      </MemoryRouter>
    </Provider>
  );

  beforeEach(() => {
    initialProps = {
      portfolioRoute: '/portfolio/foo',
      portfolio: {
        name: 'Foo'
      }
    };
  });

  it('should render correctly', () => {
    const store = mockStore({});
    const wrapper = shallow(<MemoryRouter><AddProductsToPortfolio store={ store } { ...initialProps } /></MemoryRouter>).dive();
    expect(shallowToJson(wrapper.find(AddProductsToPortfolio))).toMatchSnapshot();
  });

  it('should correctly filter service offerings', async done => {
    const store = mockStore({
      platformReducer: {
        platforms: [{ id: '1', name: 'foo' }],
        platformItems: { 1: { data: [{ id: '123', name: 'platformItem', description: 'description' }]}}
      }
    });
    apiClientMock.post(`${SOURCES_API_BASE}/graphql`, mockOnce({ body: {
      data: { application_types: [{ sources: [{ id: '1', name: 'foo' }]}]}}}));
    apiClientMock
    .get(`${TOPOLOGICAL_INVENTORY_API_BASE}/sources/1/service_offerings?filter%5Barchived_at%5D%5Bnil%5D%20=&limit=50&offset=0`,
      mockOnce({ body: { data: [], meta: { count: 123, limit: 50, offset: 123 }}}));

    let wrapper;
    await act(async () => {
      wrapper = mount(
        <ComponentWrapper store={ store }>
          <AddProductsToPortfolio { ...initialProps } />
        </ComponentWrapper>
      );
    });

    setImmediate(() => {
      const select = wrapper.find(rawComponents.Select);
      act(() => {
        select.props().onChange({ id: '1' });
      });
      wrapper.update();
      expect(wrapper.find(PlatformItem)).toHaveLength(1);
      const searchInput = wrapper.find('input').at(1);
      searchInput.getDOMNode().value = 'foo';
      searchInput.simulate('change');
      wrapper.update();
      expect(wrapper.find(PlatformItem)).toHaveLength(0);
      done();
    });
  });

  it('should check item and send correct data on submit', async done => {
    const store = mockStore({
      platformReducer: {
        platforms: [{ id: '1', name: 'foo' }],
        platformItems: { 1: { data: [{ id: '123', name: 'platformItem', description: 'description' }]}}
      }
    });
    apiClientMock.post(`${SOURCES_API_BASE}/graphql`, mockOnce({ body: {
      data: { application_types: [{ sources: [{ id: '1', name: 'foo' }]}]}}}));
    apiClientMock
    .get(`${TOPOLOGICAL_INVENTORY_API_BASE}/sources/1/service_offerings?filter%5Barchived_at%5D%5Bnil%5D%20=&limit=50&offset=0`,
      mockOnce({ body: { data: [], meta: {}}
      }));
    apiClientMock.post(`${CATALOG_API_BASE}/portfolio_items`, mockOnce({ body: { id: '999' }}));
    apiClientMock.post(`${CATALOG_API_BASE}/portfolios/321/portfolio_items`, mockOnce((req, res) => {
      expect(JSON.parse(req.body())).toEqual({ portfolio_item_id: '999' });
      return res.status(200);
    }));
    apiClientMock.get(`${CATALOG_API_BASE}/portfolios/321/portfolio_items`, mockOnce({ body: { data: []}}));

    let wrapper;
    await act(async () => {
      wrapper = mount(
        <ComponentWrapper store={ store }>
          <AddProductsToPortfolio { ...initialProps } portfolio={ { id: '321', name: 'Foo' } } />
        </ComponentWrapper>
      );
    });

    setImmediate(async () => {
      const select = wrapper.find(rawComponents.Select);
      act(() => {
        select.props().onChange({
          id: '1'
        });
      });
      wrapper.update();
      wrapper.find('input').last().simulate('change');
      await act(async () => {
        wrapper.find('button').last().simulate('click');
      });
      setImmediate(() => {
        // wait for redirect and portfolio items refresh
        expect(wrapper.find(MemoryRouter).childAt(0).props().history.location.pathname).toEqual('/portfolio/foo');
        done();
      });
    });
  });
});
