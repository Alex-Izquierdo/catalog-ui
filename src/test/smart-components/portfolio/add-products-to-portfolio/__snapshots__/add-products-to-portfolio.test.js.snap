// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AddProductsToPortfolio /> should render correctly 1`] = `
<withRouter(Connect(AddProductsToPortfolio))
  portfolio={
    Object {
      "name": "Foo",
    }
  }
  portfolioRoute="/portfolio/foo"
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
/>
`;
